net: Change the net_linkaddr struct to not use pointers

jukkar opened the pull request:
Previously the net_linkaddr struct had pointers to the link address. This is error prone and difficult to handle if cloning the packet as those pointers can point to wrong place. Mitigate this issue by allocating the space for link address in net_linkaddr struct. This will increase the size of the net_pkt by 4 bytes for IEEE 802.15.4 where the link address length is 8, but there no increase in size if link address is 6 bytes like in Ethernet/Wi-Fi.

Fixes #86652

jukkar committed:
tests: net: utils: Verify src and dst link addresses

Make sure that source and destination link addresses are correct
after cloning and pulling net_pkt.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Change the net_linkaddr struct to not use pointers

Previously the net_linkaddr struct had pointers to the link address.
This is error prone and difficult to handle if cloning the packet as
those pointers can point to wrong place. Mitigate this issue by
allocating the space for link address in net_linkaddr struct. This will
increase the size of the net_pkt by 4 octets for IEEE 802.15.4 where the
link address length is 8, but there no increase in size if link address
is 6 bytes like in Ethernet/Wi-Fi.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented:
According to the contribution guidelines, commits should not have `Fixes #XXXXX` in the body.

rlubos commented on include/zephyr/net/net_linkaddr.h:
```diff
@@ -146,6 +131,32 @@ static inline int net_linkaddr_set(struct net_linkaddr_storage *lladdr_store,
+ *
+ * @brief Copy link address from one variable to another.
+ *
+ * @param lladdr_dst The link address structure destination.
+ * @param lladdr_src The link address structure to source.
```
Param names here don't match with the actual params.

jukkar committed:
doc: net: Add link address change info to migration guide

Add information about "struct net_linkaddr" changes to 4.2
migration guide to help the developers to handle issues if
they access the net_linkaddr struct fields directly in their
code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/ip/icmpv6.c:
```diff
@@ -145,8 +145,8 @@ static int icmpv6_handle_echo_request(struct net_icmp_ctx *ctx,
 	 */
-	net_pkt_lladdr_dst(reply)->addr = NULL;
-	net_pkt_lladdr_src(reply)->addr = NULL;
+	memset(net_pkt_lladdr_dst(reply)->addr, 0, sizeof(net_pkt_lladdr_dst(reply)->addr));
```
I wonder, shouldn't we set `len` to 0 too, as now neigbor cache check seems to rely on this field?

jukkar replied:
Yes, definitely we need to set the len to 0, otherwise the address is considered valid. Will fix.

rlubos commented on subsys/net/ip/ipv6_nbr.c:
```diff
@@ -1210,7 +1212,7 @@ static int handle_ns_input(struct net_icmp_ctx *ctx,
 	const struct in6_addr *na_dst;
 	struct in6_addr *tgt;
-	struct net_linkaddr_storage src_lladdr_s;
+	struct net_linkaddr src_lladdr_s;
```
It doesn't seem we still need the second struct here? We could call `read_llao()` directly with `src_lladdr` I think?

jukkar replied:
Yes, very true, will fix.

rlubos commented on subsys/net/ip/route.c:
```diff
@@ -1079,13 +1079,15 @@ int net_route_packet(struct net_pkt *pkt, struct in6_addr *nexthop)
 	 * destination address to be the nexthop recipient.
 	 */
 	if (is_ll_addr_supported(net_pkt_iface(pkt))) {
-		net_pkt_lladdr_src(pkt)->addr = net_pkt_lladdr_if(pkt)->addr;
```
Maybe just memcpy entire struct, instead of copying each field separately? Same below.

jukkar commented:
* CI fixes
* Added info to migration guide
* I created some helpers to setup `net_linkaddr` and made the code to use those. This avoids the constant memcpy and setting the length/type of the link address

alexanderwachter commented on include/zephyr/net/net_linkaddr.h:
```diff
@@ -146,6 +131,78 @@ static inline int net_linkaddr_set(struct net_linkaddr_storage *lladdr_store,
+ * @param lladdr The link address structure.
+ * @return 0 if ok, <0 if error
+ */
+static inline int net_linkaddr_clear(struct net_linkaddr *lladdr)
```
This function could just call the create function with the src members as parameters

jukkar commented:
* Updated according to comments
* More CI fixes

jukkar commented:
* CI fixes

jukkar commented:
* More CI fixes

pdgendt commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -133,6 +133,17 @@ Bluetooth Host
 Networking
 **********
 
+* The struct :c:struct:`net_linkaddr_storage` has been renamed to struct
```
This struct doesn't exist anymore, so I don't think this directive actually works.

```suggestion
* The struct ``net_linkaddr_storage`` has been renamed to struct
```

jukkar replied:
Good point, will fix that.

pdgendt reviewed (commented):
Small comment, otherwise LGTM

jukkar commented:
* Updated the text in migration guide
